File: apps/android/app/src/main/kotlin/com/maroonedsoftware/deadair/schedule/StationClock.kt
package com.maroonedsoftware.deadair.schedule

import java.time.Duration
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Reading the station's clock, which is not this phone's.
 *
 * The API sends `YYYY-MM-DD HH:mm:ss` with no offset, deliberately: a block's ends are a READING of
 * the station's own clock rather than a moment in time, and attaching this device's timezone to one
 * would move a show that starts at eight to whatever eight o'clock means here. So every stamp is
 * parsed as a local date-time with no zone at all, displayed exactly as written, and the only
 * arithmetic done on it is the difference between two readings the station sent together.
 *
 * What that means at a clock change is worth being plain about, and it is the same thing the console
 * says: the answer is how much CLOCK is left, not how much time. On the night the clocks go back, a
 * block ending at two reads an hour shorter than it will run. That is the right answer for a screen
 * whose whole subject is what the station's clock says.
 */
private val STAMP: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")

/** A stamp as a zone-less date-time, or `null` for anything that is not one. */
fun readClock(stamp: String): LocalDateTime? =
    try {
        LocalDateTime.parse(stamp, STAMP)
    } catch (error: Exception) {
        null
    }

/** How many minutes apart two readings are. Zero when either is unreadable, which shows as "ending". */
fun minutesBetween(from: String, to: String): Long {
    val start = readClock(from) ?: return 0
    val end = readClock(to) ?: return 0
    return Duration.between(start, end).toMinutes()
}

/**
 * A number of minutes as a length somebody would say out loud.
 *
 * The unit gets coarser as the number gets bigger, which is the point of it: a block that runs once
 * a week is genuinely six days off, and "161 h" is a true answer nobody can read. Past a day the
 * minutes stop being information.
 */
fun formatSpan(minutes: Long): String {
    if (minutes <= 0) return "ending"
    if (minutes < MINUTES_PER_HOUR) return "$minutes min"

    if (minutes < MINUTES_PER_DAY) {
        val hours = minutes / MINUTES_PER_HOUR
        val rest = minutes % MINUTES_PER_HOUR
        return if (rest == 0L) "$hours h" else "$hours h $rest min"
    }

    val days = Math.round(minutes.toDouble() / MINUTES_PER_DAY)
    return if (days == 1L) "a day" else "$days days"
}

/**
 * A block's hours, with the weekday when it is not the one the station is having.
 *
 * The date is dropped rather than shown in full: these sit stacked and the useful difference between
 * them is the hour, not the year.
 */
fun formatHours(start: String, end: String, now: String): String {
    val from = readClock(start)
    val to = readClock(end)
    if (from == null || to == null) return ""

    val today = readClock(now)?.toLocalDate()
    val prefix = if (today != null && from.toLocalDate() != today) "${DAY_LABELS[from.dayOfWeek.value % 7]} " else ""
    return "$prefix${clockOf(from)}–${clockOf(to)}"
}

private fun clockOf(at: LocalDateTime): String = "%02d:%02d".format(at.hour, at.minute)

/** Sunday first, matching the API's own day numbering on a slot. */
private val DAY_LABELS = listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")

private const val MINUTES_PER_HOUR = 60L
private const val MINUTES_PER_DAY = 24L * MINUTES_PER_HOUR

File: apps/ios/Packages/DeadairCore/Sources/DeadairCore/Schedule/StationClock.swift
import Foundation

/// Reading the station's clock, which is not this phone's.
///
/// The API sends `YYYY-MM-DD HH:mm:ss` with no offset, deliberately: a block's ends are a READING
/// of the station's own clock rather than a moment in time, and attaching this device's timezone to
/// one would move a show that starts at eight to whatever eight o'clock means here. So every stamp
/// is read on a fixed calendar with no zone of its own (UTC, which has no clock changes), displayed
/// exactly as written, and the only arithmetic done on it is the difference between two readings
/// the station sent together. `apps/android`'s `StationClock.kt`, decision for decision.
///
/// At a clock change the answer is how much CLOCK is left, not how much time: on the night the
/// clocks go back, a block ending at two reads an hour shorter than it will run. That is the right
/// answer for a screen whose whole subject is what the station's clock says.
public struct StationStamp: Equatable, Sendable {
    public let year: Int
    public let month: Int
    public let day: Int
    public let hour: Int
    public let minute: Int
    public let second: Int

    /// The reading as an instant on the zone-less calendar, for subtraction and nothing else.
    var instant: Date? { stationCalendar.date(from: DateComponents(year: year, month: month, day: day, hour: hour, minute: minute, second: second)) }

    var weekday: Weekday? {
        instant.map { stationCalendar.component(.weekday, from: $0) }.flatMap(Weekday.init(rawValue:))
    }

    func sameDay(as other: StationStamp) -> Bool { year == other.year && month == other.month && day == other.day }
}

/// Gregorian, on UTC: a calendar with no daylight saving to move anything under the arithmetic.
private let stationCalendar: Calendar = {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(identifier: "UTC")!
    return calendar
}()

/// A stamp read as the station wrote it, or `nil` for anything that is not one.
public func readClock(_ stamp: String) -> StationStamp? {
    let halves = stamp.split(separator: " ")
    guard halves.count == 2 else { return nil }
    let date = halves[0].split(separator: "-").compactMap { Int($0) }
    let time = halves[1].split(separator: ":").compactMap { Int($0) }
    guard date.count == 3, time.count == 3 else { return nil }
    let reading = StationStamp(year: date[0], month: date[1], day: date[2], hour: time[0], minute: time[1], second: time[2])
    // A stamp that names a day the calendar does not have (the thirtieth of February) is not one.
    guard let instant = reading.instant, stationCalendar.component(.day, from: instant) == reading.day,
          (0..<24).contains(reading.hour), (0..<60).contains(reading.minute), (0..<60).contains(reading.second)
    else { return nil }
    return reading
}

/// How many minutes apart two readings are. Zero when either is unreadable, which shows as "ending".
public func minutesBetween(_ from: String, _ to: String) -> Int {
    guard let start = readClock(from)?.instant, let end = readClock(to)?.instant else { return 0 }
    return Int((end.timeIntervalSince(start) / 60).rounded(.towardZero))
}

/// A number of minutes as a length somebody would say out loud. Past a day the minutes stop being
/// information, so the unit gets coarser as the number grows.
public func spanOf(_ minutes: Int) -> Span {
    if minutes <= 0 { return .ending }
    if minutes < minutesPerHour { return .minutes(minutes) }
    if minutes < minutesPerDay { return .hours(minutes / minutesPerHour, minutes: minutes % minutesPerHour) }
    let days = Int((Double(minutes) / Double(minutesPerDay)).rounded())
    return days == 1 ? .aDay : .days(days)
}

/// A block's hours, with the weekday when it is not the day the station is having. `nil` for a stamp
/// that cannot be read, which is nothing to show.
public func hoursOf(start: String, end: String, now: String) -> BlockHours? {
    guard let from = readClock(start), let to = readClock(end) else { return nil }
    let today = readClock(now)
    let day = today.map { from.sameDay(as: $0) ? nil : from.weekday } ?? nil
    return BlockHours(from: Clock(hour: from.hour, minute: from.minute), to: Clock(hour: to.hour, minute: to.minute), day: day)
}

private let minutesPerHour = 60
private let minutesPerDay = 24 * 60
